server/src/main/java/org/elasticsearch/indices/recovery/RecoverySettings.java

```diff
@@ -21,16 +21,38 @@
 import org.elasticsearch.common.unit.ByteSizeUnit;
 import org.elasticsearch.common.unit.ByteSizeValue;
 import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.jdk.JavaVersion;
 import org.elasticsearch.monitor.os.OsProbe;
 import org.elasticsearch.node.NodeRoleSettings;
 
+import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
 
 public class RecoverySettings {
 
     private static final Logger logger = LogManager.getLogger(RecoverySettings.class);
 
+    private static final Boolean SNAPSHOTS_RECOVERY_FEATURE_FLAG_REGISTERED;
+
+    static {
+        final String property = System.getProperty("es.snapshots_recovery_feature_flag_registered");
+        if (Build.CURRENT.isSnapshot() && property != null) {
+            throw new IllegalArgumentException("es.snapshots_recovery_feature_flag_registered is only supported in non-snapshot builds");
```

Comment:
I'm not sure we need a feature flag for this feature, but maybe there's been some other conversation about that. I'd be happy with it simply being controlled by the `indices.recovery.use_snapshots` setting.